sdk/tests/conformance2/buffers/buffer-data-and-buffer-sub-data-sub-source.html

```diff
@@ -0,0 +1,207 @@
+
+        gl.bufferData(gl.ARRAY_BUFFER, view, gl.STATIC_DRAW, count + 1);
+        wtu.glErrorShouldBe(gl, gl.INVALID_VALUE,
+            "calling bufferData when srcOffset + length is larger than source size");
+        gl.bufferData(gl.ARRAY_BUFFER, view, gl.STATIC_DRAW, count + 1, 0);
+        wtu.glErrorShouldBe(gl, gl.INVALID_VALUE,
+            "calling bufferData when srcOffset + length is larger than source size");
+
+        gl.bufferData(gl.ARRAY_BUFFER, view, gl.STATIC_DRAW, count - 1, 2);
+        wtu.glErrorShouldBe(gl, gl.INVALID_VALUE,
+            "calling bufferData when srcOffset + length is larger than source size");
+
+        var offset = 2;
+        var size = count - offset;
+        gl.bufferData(gl.ARRAY_BUFFER, view, gl.STATIC_DRAW, offset);
+        wtu.glErrorShouldBe(gl, gl.NO_ERROR,
+            "calling bufferData with valid sub source should succeed");
+        verifyBufferData(testCases[test], 0, size, view.slice(offset, offset + size));
+
+        offset = 1;
+        size = 1;
+        gl.bufferData(gl.ARRAY_BUFFER, view, gl.STATIC_DRAW, offset, size);
+        wtu.glErrorShouldBe(gl, gl.NO_ERROR,
+            "calling bufferData with valid sub source should succeed");
+        verifyBufferData(testCases[test], 0, size, view.slice(offset, offset + size));
+
+        gl.bufferData(gl.ARRAY_BUFFER, view, gl.STATIC_DRAW, count);
+        wtu.glErrorShouldBe(gl, gl.NO_ERROR,
+            "calling bufferData with valid sub source of size 0 should succeed");
+
+        gl.deleteBuffer(buf);
+        gl.bindBuffer(gl.ARRAY_BUFFER, null);
+
+        wtu.glErrorShouldBe(gl, gl.NO_ERROR, "no GL error");
+    }
+}
+
+function bufferSubDataTest() {
+    for (var test = 0; test < testCases.length; ++test) {
+        debug("");
+        debug("Test bufferSubData with " + testCases[test].type);
+
+        var count = 4;
+        var totalBufferBytes = testCases[test].size * count * 2;
+        var buf = gl.createBuffer();
+        shouldBeNonNull(buf);
+        gl.bindBuffer(gl.ARRAY_BUFFER, buf);
+        gl.bufferData(gl.ARRAY_BUFFER, totalBufferBytes, gl.STATIC_DRAW);
+        wtu.glErrorShouldBe(gl, gl.NO_ERROR);
+
+        var arrayBuffer = new ArrayBuffer(testCases[test].size * count);
+        var view = new window[testCases[test].type](arrayBuffer);
+        for (var ii = 0; ii < count; ++ii) {
+            view[ii] = ii;
+        }
+
+        var sourceByteOffset = testCases[test].size * 2;
+
+        gl.bufferSubData(gl.ARRAY_BUFFER, sourceByteOffset, view, 0, count + 1);
+        wtu.glErrorShouldBe(gl, gl.INVALID_VALUE,
+            "calling bufferSubData when srcOffset + length is larger than source size");
+
+        gl.bufferSubData(gl.ARRAY_BUFFER, sourceByteOffset, view, count + 1);
+        wtu.glErrorShouldBe(gl, gl.INVALID_VALUE,
+            "calling bufferSubData when srcOffset + length is larger than source size");
+        gl.bufferSubData(gl.ARRAY_BUFFER, sourceByteOffset, view, count + 1, 0);
```

Comment:
Since the "optional GLuint length" argument defaults to 0, isn't this actually the same as the call above? It doesn't matter whether 0 is explicitly provided; it always means to upload the "rest" of the given view.